File: base_station/src/base_station/web/static/daq-config/node-specs.js
/** Declarative definitions for hardware-independent DAQ graph nodes. */

import {
  ENGINEERING_UNITS, booleanControl, dashboardIdentityControls, gaugeControls, numberControl,
  option, selectControl, timePlotControls,
} from './node-spec-controls.js';
import {
  finite, positive, validateDashboardIdentity, validateGauge, validatePrecision,
  validateSine, validateTimePlot,
} from './node-spec-validation.js';

export { numberControl, selectControl } from './node-spec-controls.js';


const SPECS = {
  'sine-wave': {
    category: 'Simulation',
    title: 'Sine wave',
    icon: 'icon-node-sine',
    tone: 'source',
    description: 'Generate a configurable test signal without DAQ hardware.',
    previewSource: true,
    pins: (config) => [output('signal', 'Signal', config.unit)],
    controls: (config) => [
      numberControl('amplitude', 'Amplitude', config.amplitude, config.unit),
      numberControl('periodS', 'Period', config.periodS, 's', { min: 0, step: 0.1 }),
      numberControl('offset', 'Offset', config.offset, config.unit),
      numberControl('phaseRad', 'Phase', config.phaseRad, 'rad', { step: 0.1 }),
      numberControl('randomness', 'Randomness', config.randomness, '', { min: 0, max: 1, step: 0.05 }),
      selectControl('unit', 'Unit', config.unit, ENGINEERING_UNITS.map(option)),
    ],
    badge: (config) => `${format(config.periodS)} s`,
    outputUnit: (node, pinId) => pinId === 'signal' ? node.config?.unit ?? 'V' : undefined,
    unitOutputPins: ['signal'],
    validate: validateSine,
  },
  constant: {
    category: 'Math',
    title: 'Constant',
    icon: 'icon-node-constant',
    tone: 'transform',
    description: 'Named engineering constant such as tank dry mass.',
    pins: (config) => [output('value', 'Value', config.unit)],
    controls: (config) => [
      numberControl('value', 'Value', config.value, config.unit),
      selectControl('unit', 'Unit', config.unit, ENGINEERING_UNITS.map(option)),
    ],
    badge: (config) => `${format(config.value)} ${config.unit ?? ''}`.trim(),
    outputUnit: (node, pinId) => pinId === 'value' ? node.config?.unit ?? null : undefined,
    unitOutputPins: ['value'],
    validate: (config) => finite(config.value) ? [] : ['Constant value must be finite'],
  },
  add: inferredMathSpec('Add', 'icon-node-add', 'Add two engineering signals with matching units.', 'A + B'),
  subtract: inferredMathSpec('Subtract', 'icon-node-subtract', 'Subtract one engineering signal from another.', 'A − B'),
  gain: {
    ...inferredUnarySpec('Gain', 'icon-node-gain', 'Scale an engineering signal by a dimensionless gain.', 'Scaled'),
    controls: (config) => [numberControl('gain', 'Gain', config.gain)],
    validate: (config) => finite(config.gain) ? [] : ['Gain must be finite'],
  },
  'moving-average': {
    ...inferredUnarySpec('Moving average', 'icon-node-average', 'Smooth a signal over a configurable time window.', 'Average'),
    controls: (config) => [numberControl('windowS', 'Window', config.windowS, 's', { min: 0.001, step: 0.05 })],
    validate: (config) => positive(config.windowS) ? [] : ['Moving-average window must be positive'],
  },
  'rate-of-change': {
    category: 'Math',
    title: 'Rate of change',
    icon: 'icon-node-rate',
    tone: 'transform',
    description: 'Time derivative for mass-flow and similar derived signals.',
    pins: () => [input('input', 'Signal', 'infer', '*'), output('rate', 'Rate', 'infer')],
    controls: (config) => [numberControl('windowS', 'Window', config.windowS, 's', { min: 0.01, step: 0.05 })],
    decorate(next, graph, helpers) {
      const unit = helpers.incomingUnit(next, graph, 'input');
      const outputUnit = concrete(unit) ? `${unit}/s` : 'infer';
      setPinType(next, 'input', unit ?? 'infer');
      setPinType(next, 'rate', outputUnit);
      next.badge = outputUnit;
    },
    outputUnit: (_node, pinId, resolveInput) => {
      if (pinId !== 'rate') return undefined;
      const unit = resolveInput('input');
      return concrete(unit) ? `${unit}/s` : null;
    },
    validate: (config) => positive(config.windowS) ? [] : ['Rate-of-change window must be positive'],
  },
  number: {
    category: 'Dashboard',
    title: 'Number',
    icon: 'icon-node-number',
    tone: 'result',
    description: 'Show one live engineering value on the operator dashboard.',
    pins: dashboardPins,
    controls: (config) => [
      ...dashboardIdentityControls(config),
      numberControl('precision', 'Decimals', config.precision, '', { min: 0, max: 6, step: 1 }),
      booleanControl('showUnits', 'Show units', config.showUnits),
    ],
    decorate: decorateDashboardInput,
    validate: (config) => [
      ...validateDashboardIdentity(config),
      ...validatePrecision(config.precision),
      ...(typeof config.showUnits === 'boolean' ? [] : ['Number showUnits must be on or off']),
    ],
  },
  gauge: {
    category: 'Dashboard',
    title: 'Gauge',
    icon: 'icon-node-gauge',
    tone: 'result',
    description: 'Show one live value against configured engineering limits.',
    pins: dashboardPins,
    controls: gaugeControls,
    decorate: decorateDashboardInput,
    validate: validateGauge,
  },
  'time-plot': {
    category: 'Dashboard',
    title: 'Time plot',
    icon: 'icon-node-time-plot',
    tone: 'result',
    description: 'Plot one engineering value with configurable time and value axes.',
    pins: dashboardPins,
    controls: timePlotControls,
    decorate: decorateDashboardInput,
    validate: validateTimePlot,
  },
};

let defaultsConfigured = false;

/** Configure declarative-node defaults from the authoritative server contract. */
export function configureSpecDefaults(defaults) {
  if (!defaults || typeof defaults !== 'object') throw new TypeError('DAQ node defaults are required');
  for (const [nodeType, spec] of Object.entries(SPECS)) {
    const nodeDefaults = defaults[nodeType];
    if (!nodeDefaults || typeof nodeDefaults !== 'object' || Array.isArray(nodeDefaults)) {
      throw new Error(`Missing DAQ defaults for ${nodeType}`);
    }
    spec.defaults = structuredClone(nodeDefaults);
  }
  const extras = Object.keys(defaults).filter((nodeType) => !SPECS[nodeType]);
  if (extras.length) throw new Error(`Unknown DAQ defaults: ${extras.join(', ')}`);
  defaultsConfigured = true;
}

export const SPEC_NODE_CATALOG = Object.entries(SPECS).map(([type, spec]) => ({
  type,
  category: spec.category,
  title: spec.title,
  icon: spec.icon,
  description: spec.description,
}));

export function isSpecNodeType(nodeType) {
  return Boolean(SPECS[nodeType]);
}

export function isPreviewSourceNode(node) {
  return Boolean(SPECS[node?.nodeType]?.previewSource);
}

export function createSpecNode(nodeType, common) {
  const spec = SPECS[nodeType];
  if (!spec) return null;
  const config = structuredClone(defaultsFor(nodeType, spec));
  return { ...common, tone: spec.tone, config, pins: spec.pins(config) };
}

export function decorateSpecNode(node, graph, helpers) {
  const spec = SPECS[node.nodeType];
  if (!spec) return null;
  const next = structuredClone(node);
  const config = currentConfig(defaultsFor(node.nodeType, spec), next.config ?? {});
  next.config = config;
  next.icon = spec.icon;
  next.controls = spec.controls?.(config) ?? [];
  if (spec.infer) decorateInferred(next, graph, helpers.incomingUnit, spec.infer);
  spec.decorate?.(next, graph, helpers);
  if (spec.badge) next.badge = spec.badge(config);
  return next;
}

export function validateSpecNode(node) {
  const spec = SPECS[node?.nodeType];
  return spec
    ? (spec.validate?.(currentConfig(defaultsFor(node.nodeType, spec), node.config ?? {})) ?? [])
    : null;
}

export function specOutputUnit(node, pinId, resolveInput) {
  const spec = SPECS[node?.nodeType];
  return spec?.outputUnit ? spec.outputUnit(node, pinId, resolveInput) : undefined;
}

export function patchSpecPins(node, key, value) {
  const spec = SPECS[node?.nodeType];
  if (!spec || key !== 'unit' || !spec.unitOutputPins?.length) return null;
  const targets = new Set(spec.unitOutputPins);
  return node.pins.map((pin) => targets.has(pin.id) ? { ...pin, type: String(value) } : pin);
}

function inferredMathSpec(title, icon, description, outputLabel) {
  return {
    category: 'Math', title, icon, tone: 'transform', description,
    pins: () => [input('a', 'A', 'infer', '*'), input('b', 'B', 'infer', '*'), output('result', outputLabel, 'infer')],
    infer: { inputs: ['a', 'b'], output: 'result' },
    outputUnit: (_node, pinId, resolveInput) => pinId === 'result'
      ? resolveInput('a') ?? resolveInput('b')
      : undefined,
  };
}

function inferredUnarySpec(title, icon, description, outputLabel) {
  return {
    category: 'Math', title, icon, tone: 'transform', description,
    pins: () => [input('input', 'Signal', 'infer', '*'), output('result', outputLabel, 'infer')],
    infer: { inputs: ['input'], output: 'result' },
    outputUnit: (_node, pinId, resolveInput) => pinId === 'result' ? resolveInput('input') : undefined,
  };
}

function dashboardPins() {
  return [input('value', 'Value', '*', '*')];
}

function decorateDashboardInput(next, graph, helpers) {
  const unit = helpers.incomingUnit(next, graph, 'value');
  if (concrete(unit)) setPinType(next, 'value', unit);
}

function decorateInferred(node, graph, incomingUnit, infer) {
  const unit = infer.inputs.map((pin) => incomingUnit(node, graph, pin)).find(concrete) ?? 'infer';
  for (const pin of infer.inputs) setPinType(node, pin, unit);
  setPinType(node, infer.output, unit);
  node.badge = unit;
}

function defaultsFor(nodeType, spec) {
  if (!defaultsConfigured || !spec?.defaults) {
    throw new Error(`DAQ defaults not configured before using ${nodeType}`);
  }
  return spec.defaults;
}

function currentConfig(defaults, config) {
  return Object.fromEntries(Object.entries(defaults).map(([key, fallback]) => [
    key,
    structuredClone(Object.hasOwn(config ?? {}, key) ? config[key] : fallback),
  ]));
}

function input(id, label, type, expectedType = type, options = {}) {
  return { id, label, type, expectedType, direction: 'input', kind: 'data', ...options };
}
function output(id, label, type) {
  return { id, label, type, direction: 'output', kind: 'result' };
}
function setPinType(node, pinId, unit) {
  const pin = node.pins?.find((candidate) => candidate.id === pinId);
  if (pin) pin.type = unit;
}
function concrete(unit) { return Boolean(unit && unit !== '*' && unit !== 'infer' && unit !== 'V / A'); }
function format(value) { return finite(value) ? Number(value).toLocaleString(undefined, { maximumFractionDigits: 5 }) : '—'; }

File: base_station/src/base_station/web/daq_config/node_specs.py
"""Canonical definitions for hardware-independent DAQ graph nodes."""

from __future__ import annotations

from copy import deepcopy
from math import isfinite


DASHBOARD_NODE_TYPES = frozenset({"number", "gauge", "time-plot"})
GAUGE_TYPES = {
    "dial-filled",
    "dial-needle",
    "meter-horizontal",
    "meter-vertical",
    "meter-vertical-inverted",
}

DEFAULT_CONFIGS = {
    "sine-wave": {
        "amplitude": 1,
        "periodS": 4,
        "offset": 0,
        "phaseRad": 0,
        "randomness": 0,
        "unit": "V",
    },
    "constant": {"value": 0, "unit": "kg"},
    "add": {},
    "subtract": {},
    "gain": {"gain": 1},
    "moving-average": {"windowS": 0.5},
    "rate-of-change": {"windowS": 0.5},
    "number": {
        "label": "",
        "precision": 1,
        "showUnits": True,
    },
    "gauge": {
        "label": "",
        "precision": 1,
        "type": "dial-filled",
        "showValue": True,
        "showUnits": True,
        "showRange": True,
        "min": 0,
        "low": 10,
        "high": 90,
        "max": 100,
    },
    "time-plot": {
        "label": "",
        "yScale": "auto",
        "yMin": 0,
        "yMax": 100,
    },
}

SPEC_NODE_TYPES = frozenset(DEFAULT_CONFIGS)


def normalize_spec_node(node: dict) -> bool:
    """Apply current defaults and pins to a recognized declarative node."""
    node_type = node.get("nodeType")
    if node_type not in SPEC_NODE_TYPES:
        return False
    source = node.get("config") if isinstance(node.get("config"), dict) else {}
    config = {
        key: deepcopy(source[key]) if key in source else deepcopy(default)
        for key, default in DEFAULT_CONFIGS[node_type].items()
    }
    node["config"] = config
    node["pins"] = _pins(node_type, config)
    return True


def validate_spec_node(node: dict) -> list[str] | None:
    """Return validation messages for a declarative node or None if custom."""
    node_type = node.get("nodeType")
    if node_type not in SPEC_NODE_TYPES:
        return None
    config = deepcopy(DEFAULT_CONFIGS[node_type])
    if isinstance(node.get("config"), dict):
        for key in config:
            if key in node["config"]:
                config[key] = deepcopy(node["config"][key])

    if node_type == "sine-wave":
        return _validate_sine(config)
    if node_type == "constant":
        return [] if _finite(config.get("value")) else ["Constant value must be finite"]
    if node_type == "gain":
        return [] if _finite(config.get("gain")) else ["Gain must be finite"]
    if node_type == "moving-average":
        return [] if _positive(config.get("windowS")) else ["Moving-average window must be positive"]
    if node_type == "rate-of-change":
        return [] if _positive(config.get("windowS")) else ["Rate-of-change window must be positive"]
    if node_type == "number":
        return _validate_number(config)
    if node_type == "gauge":
        return _validate_gauge(config)
    if node_type == "time-plot":
        return _validate_time_plot(config)
    return []


def _pins(node_type: str, config: dict) -> list[dict]:
    if node_type == "sine-wave":
        return [_output("signal", "Signal", str(config.get("unit", "V")))]
    if node_type == "constant":
        return [_output("value", "Value", str(config.get("unit", "kg")))]
    if node_type in {"add", "subtract"}:
        label = "A + B" if node_type == "add" else "A − B"
        return [
            _input("a", "A", "infer", "*"),
            _input("b", "B", "infer", "*"),
            _output("result", label, "infer"),
        ]
    if node_type in {"gain", "moving-average"}:
        label = "Scaled" if node_type == "gain" else "Average"
        return [_input("input", "Signal", "infer", "*"), _output("result", label, "infer")]
    if node_type == "rate-of-change":
        return [_input("input", "Signal", "infer", "*"), _output("rate", "Rate", "infer")]
    return [_input("value", "Value", "*", "*")]


def _validate_sine(config: dict) -> list[str]:
    issues: list[str] = []
    for key, label in (
        ("amplitude", "amplitude"),
        ("periodS", "period"),
        ("offset", "offset"),
        ("phaseRad", "phase"),
        ("randomness", "randomness"),
    ):
        if not _finite(config.get(key)):
            issues.append(f"Sine-wave {label} must be finite")
    if _finite(config.get("periodS")) and float(config["periodS"]) < 0:
        issues.append("Sine-wave period cannot be negative")
    if _finite(config.get("randomness")) and not 0 <= float(config["randomness"]) <= 1:
        issues.append("Sine-wave randomness must be between 0 and 1")
    if not str(config.get("unit", "")).strip():
        issues.append("Sine-wave unit is required")
    return issues


def _validate_dashboard_identity(config: dict) -> list[str]:
    issues: list[str] = []
    if not str(config.get("label", "")).strip():
        issues.append("Dashboard widget requires a label")
    return issues


def _validate_precision(value: object) -> list[str]:
    if isinstance(value, bool) or not isinstance(value, int) or not 0 <= value <= 6:
        return ["Dashboard decimal places must be 0 through 6"]
    return []


def _validate_number(config: dict) -> list[str]:
    issues = [*_validate_dashboard_identity(config), *_validate_precision(config.get("precision"))]
    if not isinstance(config.get("showUnits"), bool):
        issues.append("Number showUnits must be on or off")
    return issues


def _validate_gauge(config: dict) -> list[str]:
    issues = [*_validate_dashboard_identity(config), *_validate_precision(config.get("precision"))]
    if config.get("type") not in GAUGE_TYPES:
        issues.append("Select a supported dashboard gauge type")
    for key in ("showValue", "showUnits", "showRange"):
        if not isinstance(config.get(key), bool):
            issues.append(f"Gauge {key} must be on or off")

    minimum = config.get("min")
    maximum = config.get("max")
    if not _finite(minimum) or not _finite(maximum) or float(maximum) <= float(minimum):
        issues.append("Gauge maximum must be greater than minimum")
        return issues
    low = config.get("low")
    high = config.get("high")
    if not _optional_limit(low, float(minimum), float(maximum), high_side=False):
        issues.append("Gauge low limit must be within the display range")
    if not _optional_limit(high, float(minimum), float(maximum), high_side=True):
        issues.append("Gauge high limit must be within the display range")
    if _finite(low) and _finite(high) and float(low) > float(high):
        issues.append("Gauge low limit cannot exceed the high limit")
    return issues


def _validate_time_plot(config: dict) -> list[str]:
    issues = _validate_dashboard_identity(config)
    if config.get("yScale") not in {"auto", "fixed"}:
        issues.append("Time-plot Y range must be Auto or Fixed")
    if config.get("yScale") == "fixed":
        minimum = config.get("yMin")
        maximum = config.get("yMax")
        if not _finite(minimum) or not _finite(maximum) or float(maximum) <= float(minimum):
            issues.append("Time-plot Y maximum must be greater than Y minimum")
    return issues


def _optional_limit(value: object, minimum: float, maximum: float, *, high_side: bool) -> bool:
    if value in (None, ""):
        return True
    if not _finite(value):
        return False
    number = float(value)
    return minimum < number <= maximum if high_side else minimum <= number < maximum


def _input(pin_id: str, label: str, pin_type: str, expected: str | None = None) -> dict:
    return {
        "id": pin_id,
        "label": label,
        "type": pin_type,
        "expectedType": expected or pin_type,
        "direction": "input",
        "kind": "data",
    }


def _output(pin_id: str, label: str, pin_type: str) -> dict:
    return {"id": pin_id, "label": label, "type": pin_type, "direction": "output", "kind": "result"}


def _finite(value: object) -> bool:
    return isinstance(value, (int, float)) and not isinstance(value, bool) and isfinite(float(value))


def _positive(value: object) -> bool:
    return _finite(value) and float(value) > 0
